### PR TITLE
Add decimal support for floor function

### PR DESCRIPTION
**Pull request by pramodsatya**

### CODE REVIEW

**Review by karteekmurthys — APPROVED** [review]
Thanks Pramod for adding these changes!

**Comment by karteekmurthys** [comment]
@pramodsatya the unit tests have failed. There seems to be a fix: https://github.com/facebookincubator/velox/pull/8761.
Please rebase.
```
[  SKIPPED ] HashTableTests/HashTableTest.offsetOverflowLoadTags/0
[  SKIPPED ] HashTableTests/HashTableTest.offsetOverflowLoadTags/1
[  FAILED  ] 1 test, listed below:
[  FAILED  ] MultiFragmentTest.partitionedOutputWithLargeInput

```

File: velox/functions/prestosql/DecimalVectorFunctions.cpp [hunk 1]
```diff
@@ -451,6 +451,36 @@ class Round {
...
+ public:
+  template <typename R, typename A>
+  inline static void apply(
+      R& r,
+      const A& a,
+      uint8_t /*aPrecision*/,
+      uint8_t aRescale,
+      const ExtraParams& /*param*/) {
+    auto rescaleFactor = DecimalUtil::kPowersOfTen[aRescale];
+    auto increment = (a % rescaleFactor) < 0 ? -1 : 0;
```
Comment (majetideepak) [inline velox/functions/prestosql/DecimalVectorFunctions.cpp:464]: `auto increment = (a < 0) ? -1 : 0;` should do?

[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:466]: Thanks for the suggestion, added a comment. This is borrowed from the following function implementation in Presto hence `% rescaleFactor`:
```
long rescaleFactor = Decimals.longTenToNth((int) numScale);
long increment = (num % rescaleFactor) < 0 ? -1 : 0;
```

Comment (pramodsatya) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:466]: Thanks for the suggestion, modifying it to `(a < 0)` results in different results from Presto for cases where `a % rescaleFactor` is `0`. The increment calculation is borrowed from Presto:
```
long rescaleFactor = Decimals.longTenToNth((int) numScale);
long increment = (num % rescaleFactor) < 0 ? -1 : 0;
```

The following query for instance returns `-124` in velox with `(a < 0)` but returns `-123` in Presto:
```
presto> select floor(CAST(DECIMAL '-123.00' AS DECIMAL(5,2)));
 _col0
-------
 -123
```
Could you please share how to proceed?

File: velox/functions/prestosql/DecimalVectorFunctions.cpp [hunk 2]
```diff
@@ -451,6 +451,36 @@ class Round {
...
+      uint8_t /*rScale*/) {
+    return fromScale;
+  }
+
+  inline static std::pair<uint8_t, uint8_t> computeResultPrecisionScale(
+      const uint8_t aPrecision,
+      const uint8_t aScale,
+      const ExtraParams& /*extraParams*/) {
+    return {
+        std::min(38, aPrecision - aScale + std::min((uint8_t)1, aScale)), 0};
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:483]: Presto calculates the result precision as follows and the same is borrowed here:
```
@Constraint(variable = "rp", expression = "p - s + min(s, 1)")
```
Added a comment indicating the same, thanks.

File: velox/functions/prestosql/DecimalVectorFunctions.cpp (hunk 1 above)
Comment (majetideepak) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:466]: Thanks!

[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:466]: > Also, why % rescaleFactor ? Its a bit odd to see % return -ve number.

@aditi-pandit that is the semantics of modulo of negative values in most cases.

**Review by majetideepak — APPROVED** [review]
Thanks, @pramodsatya

**Comment by pramodsatya** [comment]
@Yuhta could you please help review/merge these changes? Thanks.

File: velox/functions/prestosql/DecimalVectorFunctions.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/functions/prestosql/DecimalVectorFunctions.cpp:466]: @majetideepak : I see. Thanks for explaining.

**Comment by Yuhta** [comment]
@pramodsatya Have we run the fuzzer manually on `floor` for an extended period with this addition?

**Comment by pramodsatya** [comment]
> @pramodsatya Have we run the fuzzer manually on `floor` for an extended period with this addition?
> 
> Hi @Yuhta, no I haven't tested this addition with the fuzzer manually, will look into how to do this and share the findings soon. Thanks.

Hi @Yuhta, I tested this change with the fuzzer manually for ~90mins and did not see any errors. Could you please share if additional testing/changes are needed? Thanks

**Pull request closed without merging**